syntax = "proto3";

package banking;

service MarketplaceService {
  // Product management
  rpc GetProduct (GetProductRequest) returns (Product);
  rpc ListProducts (ListProductsRequest) returns (ListProductsResponse);
  rpc CreateProduct (CreateProductRequest) returns (Product);
  rpc UpdateProduct (UpdateProductRequest) returns (Product);
  rpc DeleteProduct (DeleteProductRequest) returns (DeleteProductResponse);
  
  // Order management
  rpc CreateOrder (CreateOrderRequest) returns (Order);
  rpc GetOrder (GetOrderRequest) returns (Order);
  rpc ListOrders (ListOrdersRequest) returns (ListOrdersResponse);
  rpc UpdateOrderStatus (UpdateOrderStatusRequest) returns (Order);
}

message GetProductRequest {
  string product_id = 1;
}

message ListProductsRequest {
  string category = 1;
  bool active_only = 2;
  int32 page = 3;
  int32 page_size = 4;
}

message ListProductsResponse {
  repeated Product products = 1;
  int32 total_count = 2;
}

message CreateProductRequest {
  string name = 1;
  string description = 2;
  string price = 3;
  string category = 4;
  string image_url = 5;
  string created_by = 6;
}

message UpdateProductRequest {
  string product_id = 1;
  string name = 2;
  string description = 3;
  string price = 4;
  string category = 5;
  string image_url = 6;
  bool is_active = 7;
}

message DeleteProductRequest {
  string product_id = 1;
}

message DeleteProductResponse {
  bool success = 1;
  string message = 2;
}

message CreateOrderRequest {
  string user_id = 1;
  repeated OrderItemRequest items = 2;
  string payment_method = 3;
}

message OrderItemRequest {
  string product_id = 1;
  int32 quantity = 2;
}

message GetOrderRequest {
  string order_id = 1;
}

message ListOrdersRequest {
  string user_id = 1;
  string status = 2;
  string start_date = 3;
  string end_date = 4;
  int32 page = 5;
  int32 page_size = 6;
}

message ListOrdersResponse {
  repeated Order orders = 1;
  int32 total_count = 2;
}

message UpdateOrderStatusRequest {
  string order_id = 1;
  string status = 2;
}

message Product {
  string product_id = 1;
  string name = 2;
  string description = 3;
  string price = 4;
  string category = 5;
  string created_by = 6;
  string date_created = 7;
  bool is_active = 8;
  string image_url = 9;
}

message OrderItem {
  string item_id = 1;
  string product_id = 2;
  string product_name = 3;
  int32 quantity = 4;
  string price = 5;
  string subtotal = 6;
}

message Order {
  string order_id = 1;
  string user_id = 2;
  string total_amount = 3;
  string status = 4;
  string date_created = 5;
  string date_updated = 6;
  string payment_method = 7;
  string reference_transaction = 8;
  repeated OrderItem items = 9;
}